Use new common time dependent options in Sphere and BCO domains

knelli2 opened the pull request:
## Proposed changes

Final PR in the series that revamps the time dependent options for the Sphere and BCO domains.

This actually adds all the new common options to the domain creators.

Fixes #5584.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
The options for the Expansion and Rotation map in the BinaryCompactObject and Sphere domains have changed slightly. In the BinaryCompactObject, they are now
```yaml
    TimeDependentMaps:
      ...
      ExpansionMap:
        InitialValues: [1.0, -4.6148457646200002e-05, 0.0] # <-- Must specify all derivs
        AsymptoticVelocityOuterBoundary: -1.0e-6 # <-- Same as before
        DecayTimescaleOuterBoundary: 50.0 # <-- Different name, same value as before
```

In the Sphere domain, you can optionally specify the same options as the BinaryCompactObject like above to use functions of time that expire. Additionally, you can specify options to use SettleToConstant functions of time in the Sphere domain with the options below
```yaml
    TimeDependentMaps:
      ...
      ExpansionMap:
        InitialValues: [1.0, -4.6148457646200002e-05, 0.0] # <-- Must specify all derivs
        InitialValuesOuterBoundary: [1.0, 0.0, 0.0] # <-- Must specify all derivs
        DecayTimescale: 52.0 # <-- Different name
        DecayTimescaleOuterBoundary: 50.0 # <-- Different name
      RotationMap:
        InitialQuaternions: [[1.0, 0.0, 0.0, 0.0]] # <-- Need quat. Can optionally specify derivs
        DecayTimescale: 50 # <-- Different name
```

Additionally, for any of the time dependent map options, you can choose the function of time from a volume H5 file by specifying the following options (not just for Expansion, but for Rotation and Translation as well and shape is slightly different)

```yaml
    TimeDependentMaps:
      ...
      ExpansionMap:
        H5Filename: BbhVolume0.h5
        SubfileName: VolumeData
      ShapeMapA:
        H5Filename: BbhVolume0.h5
        SubfileName: VolumeData
        TransitionEndsAtCube: true
```


<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

~Depends on and includes #6109, #6113, and #6114.~

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Allow FoTs to create themselves at specific times

nilsvu commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:
```diff
@@ -69,6 +69,19 @@ std::unique_ptr<FunctionOfTime> PiecewisePolynomial<MaxDeriv>::get_clone()
+std::unique_ptr<FunctionOfTime> PiecewisePolynomial<MaxDeriv>::create_at_time(
+    const double t, const double expiration_time) const {
+  std::vector<DataVector> all_derivs_vec = func_and_all_derivs(t);
+  std::array<DataVector, MaxDeriv + 1> all_derivs{};
+  for (size_t i = 0; i < all_derivs.size(); i++) {
+    gsl::at(all_derivs, i) = std::move(all_derivs_vec[i]);
+  }
```
I think this is `make_array<MaxDeriv+1>(func_and_all_derivs(t))`

nilsvu commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -79,6 +79,21 @@ std::unique_ptr<FunctionOfTime> QuaternionFunctionOfTime<MaxDeriv>::get_clone()
+QuaternionFunctionOfTime<MaxDeriv>::create_at_time(
+    const double t, const double expiration_time) const {
+  std::vector<DataVector> all_angle_derivs_vec = angle_func_and_all_derivs(t);
+  std::array<DataVector, MaxDeriv + 1> all_angle_derivs{};
+  for (size_t i = 0; i < all_angle_derivs.size(); i++) {
+    gsl::at(all_angle_derivs, i) = std::move(all_angle_derivs_vec[i]);
+  }
```
also `make_array`

nilsvu commented on src/Domain/Creators/TimeDependentOptions/TranslationMap.hpp:
```diff
@@ -22,29 +22,42 @@ namespace domain::creators::time_dependent_options {
-struct TranslationMapOptions {
-  using type = Options::Auto<TranslationMapOptions, Options::AutoLabel::None>;
-  static std::string name() { return "TranslationMap"; }
+struct TranslationMapHardcodedOptions {
```
I always read "hardcore options" 💪

Suggestion: keep this class named `TranslationMapOptions`. And the tag below name just `TranslationMap` because it's already in the `time_dependent_options` namespace.

nilsvu reviewed (changes requested):
ok I'll need more coffee for this one

You can squash everything into one commit, except keep the "Allow FoTs to create themselves" commit separate.

knelli2 committed:
Allow option tag test to work with Auto

knelli2 committed:
Use common time dep opts in BCO and sphere domains

Also this moves the FromVolumeFile option up a level in options
so all the time dependent options are edited as well

nilsvu commented on tests/Unit/Domain/FunctionsOfTime/Test_IntegratedFunctionOfTime.cpp:
```diff
@@ -59,6 +59,21 @@ void test(const gsl::not_null<FunctionsOfTime::FunctionOfTime*> f_of_t,
+
+  CAPTURE(rotation);
+  //   const double check_time = 0.5 * (new_time + new_expiration);
+  const double check_time = new_time;
```
Want to check here that the FoT is also the same a bit later?

knelli2 replied:
An IntegratedFunctionOfTime is only valid at an update time, so we can't check at a later time. But I will remove the commented out line

knelli2 commented:
@nilsvu Pushed my changes. This should be ready for review. Good luck 😉

nilsvu commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:
```diff
@@ -69,6 +69,14 @@ std::unique_ptr<FunctionOfTime> PiecewisePolynomial<MaxDeriv>::get_clone()
+std::unique_ptr<FunctionOfTime> PiecewisePolynomial<MaxDeriv>::create_at_time(
+    const double t, const double expiration_time) const {
+  return std::make_unique<PiecewisePolynomial>(
+      t, make_array<DataVector, MaxDeriv + 1>(func_and_all_derivs(t)),
```
Can you use `func_and_derivs(t)` so you don't need the `make_array` and elide the array->vector->array transformation?

nilsvu commented on src/Domain/FunctionsOfTime/QuaternionFunctionOfTime.cpp:
```diff
@@ -79,6 +80,16 @@ std::unique_ptr<FunctionOfTime> QuaternionFunctionOfTime<MaxDeriv>::get_clone()
+    const double t, const double expiration_time) const {
+  return std::make_unique<QuaternionFunctionOfTime>(
+      t, func(t),
+      make_array<DataVector, MaxDeriv + 1>(angle_func_and_all_derivs(t)),
```
Also here `angle_f_of_t_.func_and_derivs(t)`

knelli2 replied:
`func_and_derivs` is private in the PiecewisePolynomial

nilsvu replied:
could just make it public

nilsvu commented on src/Domain/Creators/TimeDependentOptions/BinaryCompactObject.hpp:
```diff
@@ -146,105 +148,30 @@ struct TimeDependentMapOptions {
-      using type = std::array<double, 3>;
-      static constexpr Options::String help = {"The initial angular velocity."};
-    };
+  using ExpansionMapOptions =
```
I got confused here because `ExpansionMapOptions` is also a struct in Expansion.hpp, but not the one you are referring to here.

nilsvu commented on src/Domain/Creators/TimeDependentOptions/ExpansionMap.hpp:
```diff
@@ -21,88 +24,110 @@ namespace domain::creators::time_dependent_options {
 struct ExpansionMapOptions {
-  using type = Options::Auto<ExpansionMapOptions, Options::AutoLabel::None>;
-  static std::string name() { return "ExpansionMap"; }
+  using type = ExpansionMapOptions;
```
If you change this line to the `Auto<variant<>>` then you don't need the extra option tag below and the confusion in the other comment is resolved. Or delete this line.

nilsvu commented on src/Domain/Creators/TimeDependentOptions/ExpansionMap.cpp:
```diff
@@ -4,87 +4,206 @@
+    }
+
+    if (pp_volume_fot != nullptr) {
+      result =
```
(optional) you can just return in these code branches

nilsvu commented on src/Domain/Creators/TimeDependentOptions/ExpansionMap.cpp:
```diff
@@ -4,87 +4,206 @@
+    const auto& hard_coded_options =
+        std::get<ExpansionMapOptions<AllowSettleFoTs>>(expansion_map_options);
+
+    if (hard_coded_options.asymptotic_velocity_outer_boundary.has_value()) {
```
confused that `asymptotic_velocity_outer_boundary` is not actually used below. Add a comment?

nilsvu commented on src/Domain/Creators/TimeDependentOptions/ExpansionMap.cpp:
```diff
@@ -4,87 +4,206 @@
+get_expansion_outer_boundary(
+    const std::variant<ExpansionMapOptions<AllowSettleFoTs>, FromVolumeFile>&
+        expansion_map_options,
+    const double initial_time) {
```
(optional) this function could be merged into the one above with a boolean template param or a runtime boolean argument. Maybe that would make it a bit easier to understand (up to you).

nilsvu commented on src/Domain/Creators/TimeDependentOptions/ExpansionMap.cpp:
```diff
@@ -4,87 +4,206 @@
+           "SettleToConstant, but SettleToConstant functions of time aren't "
+           "allowed.");
+
+    result = volume_fot.at(name)->get_clone();
```
(optional) I think you can always just return `create_at_time()` because it defaults to `get_clone()`, then you don't have to decide if the FoT actually does something special in `create_at_time`.

nilsvu commented on src/Domain/Creators/TimeDependentOptions/ShapeMap.hpp:
```diff
@@ -189,11 +223,9 @@ struct YlmsFromSpEC {
   static constexpr Options::String help = {
       "Options for a time-dependent distortion (shape) map about the "
-      "specified object. Specify 'None' to not use this map."};
+      "specified object.."};
```
remove one `.`

nilsvu commented on src/Domain/Creators/TimeDependentOptions/BinaryCompactObject.cpp:
```diff
@@ -189,102 +206,56 @@ TimeDependentMapOptions<IsCylindrical>::create_functions_of_time(
-  // actually use the angles themselves. We only use their derivatives
-  // (omega) to determine map parameters. In theory we could determine
-  // each initial angle from the input axis-angle representation, but
-  // we don't need to.
```
Could you add some docs to `QuaternionFunctionOfTime` so we don't get confused by this again?

knelli2 committed:
Add docs to QuaternionFunctionOfTime
